repose-aggregator/src/docs/asciidoc/filters/keystone-v2.adoc

```diff
@@ -1,52 +1,413 @@
+A list of Rackspace Cloud Feeds endpoints for Identity Events can be found at
+https://one.rackspace.com/display/auth/Identity+Endpoints#IdentityEndpoints-EndpointsConsumed[the internal Rackspace Wiki page linked here].
+====
 
-== Additional Information
-*Remove this section if not needed.*
+[source,xml]
+.keystone-v2.cfg.xml
+----
+<?xml version="1.0" encoding="UTF-8"?>
+<keystone-v2 xmlns="http://docs.openrepose.org/repose/keystone-v2/v1.0">
+    <identity-service uri="http://identity.example.com"/>
+    <cache>
+        <atom-feed id="some-feed"/> <!--1-->
+    </cache>
+</keystone-v2>
+----
+<1> The unique ID of a feed defined in the <<../services/atom-feed-consumption.adoc#, Atom Feed Consumption service>> configuration.
+
+=== Tenant ID Validation
+Tenant ID Validation is the capability of this filter to parse a tenant ID out of a request URI and validate it against the tenant ID(s) available in the response token from the Keystone v2 Identity service.
+
+[source,xml]
+.keystone-v2.cfg.xml
+----
+<?xml version="1.0" encoding="UTF-8"?>
+<keystone-v2 xmlns="http://docs.openrepose.org/repose/keystone-v2/v1.0">
+    <identity-service uri="http://identity.example.com"/>
+    <tenant-handling send-all-tenant-ids="false"> <!--1-->
+        <validate-tenant strip-token-tenant-prefixes="/foo:/bar-" <!--2--> <!--3-->
+                         enable-legacy-roles-mode="false" <!--4-->
+        >
+            <uri-extraction-regex>${your-regex}</uri-extraction-regex> <!--5-->
+        </validate-tenant>
+        <send-tenant-id-quality default-tenant-quality="0.9" <!--6--> <!--7-->
+                                uri-tenant-quality="0.7" <!--8-->
+                                roles-tenant-quality="0.5" <!--9-->
+        >
+    </tenant-handling>
+</keystone-v2>
+----
+<1> Indicates if all the Tenant IDs from the user and the roles the user has should be sent or not. (Default: false)
+<2> If this element is included, then Tenant ID Validation will be enforced based on the value extracted from the URI.
+<3> A `/` delimited list of prefixes to attempt to strip from the Tenant ID in the token response from the Keystone v2 Identity service.
+    The post-strip Tenant ID is only used in the Tenant Validation check.
+<4> If in legacy roles mode, then all roles associated with a user token are forwarded.
+    If NOT in legacy roles mode, then roles which aren't tied to the tenant provided in the request will NOT be forwarded UNLESS the user has a pre-authorized role. (Default: false)
+<5> The https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html[Java Regular Expression] with a single capturing group around the portion of the URI to extract the Tenant ID from for validation.
+<6> If this element is included, then include Quality parameters on all the tenant ID headers sent.
+<7> The default tenant has the highest quality. (Default: 0.9)
+<8> Followed by the one that matches the URI (if any). (Default: 0.7)
```

Comment:
Basically here is the logic from:

 - https://github.com/rackerlabs/repose/blob/927e229ae3b727f00c12c6ecd4ae0251ce1b6b48/repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala#L454-L487

```
IF Sending All Tenants AND Send Quality, THEN add Default Tenant at the Default Quality AND the URI Tenant at the URI Quality (if present).
ELSE IF Sending All Tenants (AND NOT Send Quality), THEN send all of the tenants without a quality
ELSE IF Sending Quality (AND NOT All Tenants), THEN ...
   IF Default Tenant and URI Tenant, THEN send the URI Tenant with the highest quality between the URI and Default.
   ELSE IF the URI Tenant (AND NOT Default Tenant), THEN send the URI Tenant with URI Quality.
   ELSE IF the Default Tenant (AND NOT URI Tenant), THEN send the Default Tenant with Default Quality.
   ELSE IF the Role Tenants, THEN send the first Role Tenant and the Role Quality.
   ELSE NONE.
ELSE ...
   IF the URI Tenant, THEN send the URI Tenant without Quality.
   ELSE IF the Default Tenant, THEN send the Default Tenant without Quality.
   ELSE IF the Role Tenants, THEN send the first Role Tenant without Quality.
   ELSE NONE.
```